data-resource/src/lib.rs

```diff
@@ -15,48 +15,31 @@ use data_error::Result;
 use serde::Serialize;
 use std::{fmt::Debug, hash::Hash, path::Path};
 
-// The `ResourceId` type is a wrapper around the hash value of the resource.
-//
-// To export another `ResourceId` type as the default for cryptographic or non-cryptographic hash,
-// implement the `ResourceIdTrait` trait for your type and re-export it here behind the
-// right feature flag.
+#[cfg(feature = "hash_blake3")]
+pub mod blake3;
 
-#[cfg(not(feature = "non-cryptographic-hash"))]
-mod blake3;
-#[cfg(not(feature = "non-cryptographic-hash"))]
-pub use blake3::ResourceId as Resource;
-#[cfg(not(feature = "non-cryptographic-hash"))]
-pub type ResourceId = <Resource as ResourceIdTrait>::HashType;
-
-#[cfg(feature = "non-cryptographic-hash")]
-mod crc32;
-#[cfg(feature = "non-cryptographic-hash")]
-pub use crc32::ResourceId as Resource;
-#[cfg(feature = "non-cryptographic-hash")]
-pub type ResourceId = <Resource as ResourceIdTrait>::HashType;
+#[cfg(feature = "hash_crc32")]
+pub mod crc32;
 
 /// This trait defines a generic type representing a resource identifier.
 ///
 /// Resources are identified by a hash value, which is computed from the resource's data.
 /// The hash value is used to uniquely identify the resource.
 ///
 /// Implementors of this trait must provide a way to compute the hash value from the resource's data.
-pub trait ResourceIdTrait {
-    /// Associated type representing the hash used by this resource identifier.
-    type HashType: Debug
-        + Display
-        + FromStr
-        + Clone
-        + PartialEq
-        + Eq
-        + Ord
-        + PartialOrd
-        + Hash
-        + Serialize;
-
+pub trait ResourceId: Debug
++ Display //todo: I guess this chain of traits can be coded in a nicer way
```

Comment:
In this case, probably we should check if each of these traits really needed.

Reply:
Yes
Here's why each of these traits is important:
- `Debug`: helps with debugging and because structs containing the hash will need to implement Debug.
- `FromStr` and `Display`: needed for parsing, and printing the hash.
- `PartialEq`, `Eq`, `Ord`, `PartialOrd`: comparing and ordering hashes.
- `Hash`: for using the hash type in collections like `HashMap`.

The only trait we do not need at the moment is `Serialize`. I included it because I plan to implement custom serialization for `ResourceIndex` later.



---

Do you think it's a good idea to add comments explaining why each trait is necessary for developers who work on the project later?

Reply:
It seems we could take one of these approaches here:
* "Opinionated" approach, where we fix all useful traits in the very core. This can save effort but might stand in the way sometime.
* "Flexible" approach, where we allow developers combine `ResourceIdTrait` with other traits explicitly. E.g. declaring `Id: ResourceIdTrait + Hash` when they need hash tables like in `fs-index`. Not always users of the framework would need `fs-index`, so maybe not always they would need `Hash` trait.

In fact, we probably should find balance between these two approaches.